walkeeper/src/receive_wal.rs

```diff
@@ -128,19 +83,27 @@ impl<'pg> ReceiveWalConn<'pg> {
         }
 
         // if requested, ask pageserver to fetch wal from us
-        // xxx: this place seems not really fitting
-        if swh.conf.pageserver_addr.is_some() {
-            // Need to establish replication channel with page server.
-            // Add far as replication in postgres is initiated by receiver, we should use callme mechanism
-            let conf = swh.conf.clone();
-            let timelineid = swh.timeline.get().timelineid;
-            let _ = thread::Builder::new()
-                .name("request_callback thread".into())
-                .spawn(move || {
-                    request_callback(conf, timelineid, tenant_id);
+        // as long as this wal_stream is alive, callmemaybe thread
+        // will send requests to pageserver
+        let _guard = match swh.conf.pageserver_addr {
+            Some(_) => {
+                // Need to establish replication channel with page server.
+                // Add far as replication in postgres is initiated by receiver, we should use callme mechanism
+                let timelineid = swh.timeline.get().timelineid;
+                let tx_clone = swh.tx.clone();
+                swh.tx
+                    .blocking_send(CallmeEvent::Subscribe(tenant_id, timelineid))
```

Comment:
You can use unbounded channel and then no blocking is needed

Reply:
Why would we need it if we send the message from a sync code anyway?